### PR TITLE
Replace storageimpl type with scalar_type and backend

### PR DESCRIPTION
**Pull request by cpuhrsch**

### CODE REVIEW

File: torch/csrc/DynamicTypes.cpp [hunk 1]
```diff
@@ -69,7 +69,11 @@ at::Type* get_type(const std::string& name, bool is_cuda, bool is_sparse) {
 
 PyTypeObject* getPyTypeObject(const at::Storage& storage)
 {
-  auto it = attype_to_py_storage_type.find(&storage.pImpl()->type());
+  auto attype = get_type(
+      at::toString((storage.pImpl())->scalar_type()),
```
Comment (ezyang) [inline torch/csrc/DynamicTypes.cpp]: This looks really weird. Why are you calling `at::toString` here?

Comment (cpuhrsch) [reply torch/csrc/DynamicTypes.cpp]: Because get_type needs a name to look up the correct type.

File: aten/src/ATen/StorageImpl.cpp [hunk 2]
```diff
@@ -28,11 +28,11 @@ StorageImpl::StorageImpl(
           allocator,
           resizable) {}
 
-Type& StorageImpl::type() {
-  if (data_ptr.device().is_cuda()) {
-    return globalContext().getType(Backend::CUDA, scalar_type);
+Backend get_backend(StorageImpl* storage_impl) {
```
Comment (ezyang) [inline aten/src/ATen/StorageImpl.cpp]: With the eventual plan to move it out of this file, right? :)

File: torch/csrc/DynamicTypes.cpp [hunk 3]
```diff
@@ -69,7 +69,11 @@ at::Type* get_type(const std::string& name, bool is_cuda, bool is_sparse) {
 
 PyTypeObject* getPyTypeObject(const at::Storage& storage)
 {
-  auto it = attype_to_py_storage_type.find(&storage.pImpl()->type());
+  auto attype = get_type(
```
Comment (ezyang) [inline torch/csrc/DynamicTypes.cpp]: use globalContext().getType() and then you don't need to convert the scalar type to string.

File: aten/src/ATen/StorageImpl.h [hunk 4]
```diff
@@ -112,4 +110,7 @@ struct AT_API StorageImpl : public Retainable {
   }
 };
 
+namespace detail {
+AT_API Backend get_backend(StorageImpl* storage_impl);
+}
```
Comment (ezyang) [inline aten/src/ATen/StorageImpl.h]: NB: at some point we're going to have to move this out of this file. (Doesn't have to be this PR).

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/StorageImpl.cpp [hunk 5]
```diff
@@ -28,11 +28,12 @@ StorageImpl::StorageImpl(
           allocator,
           resizable) {}
 
-Type& StorageImpl::type() {
-  if (data_ptr.device().is_cuda()) {
-    return globalContext().getType(Backend::CUDA, scalar_type);
+namespace detail {
+Backend get_backend(StorageImpl* storage_impl) {
```
Comment (gchanan) [inline aten/src/ATen/StorageImpl.cpp]: I don't think this really addresses the issue.  The point is that there is not a one-to-one function from storage to backend -- don't write a function like this.  Why doesn't the usage in `checked_cast_storage` just take a DeviceType or something?  And the usage in getPyTypeObject you can just call get_backend in that same file.

**Pull request closed without merging**